clients/adapter-kafka/src/main/java/org/eclipse/hono/adapter/client/command/kafka/KafkaBasedCommand.java

```diff
@@ -45,99 +44,120 @@
         }
-        final String originalDeviceId = getHeader(record, MessageHelper.APP_PROPERTY_DEVICE_ID);
-        if (Strings.isNullOrEmpty(originalDeviceId)) {
-            validationErrorJoiner.add("device identifier is not set");
-        } else if (!originalDeviceId.equals(record.key())) {
-            validationErrorJoiner.add("device identifier not set as record key");
+        final String tenantId = honoTopic.getTenantId();
+        final String deviceId = getHeader(record, MessageHelper.APP_PROPERTY_DEVICE_ID);
+        if (Strings.isNullOrEmpty(deviceId)) {
+            throw new IllegalArgumentException("device identifier is not set");
+        } else if (!deviceId.equals(record.key())) {
+            throw new IllegalArgumentException("device identifier not set as record key");
         }
+
+        String validationError = null;
         final String subject = getHeader(record, MessageHelper.SYS_PROPERTY_SUBJECT);
         if (subject == null) {
-            validationErrorJoiner.add("subject not set");
+            validationError = "subject not set";
         }
         final String contentType = getHeader(record, MessageHelper.SYS_PROPERTY_CONTENT_TYPE);
         final String correlationId = getHeader(record, MessageHelper.SYS_PROPERTY_CORRELATION_ID);
 
         return new KafkaBasedCommand(
-                validationErrorJoiner.length() > 0 ? Optional.of(validationErrorJoiner.toString()) : Optional.empty(),
+                Optional.ofNullable(validationError),
                 record,
                 tenantId,
                 deviceId,
-                originalDeviceId,
                 Strings.isNullOrEmpty(correlationId) ? null : correlationId,
                 subject,
                 contentType);
     }
 
+    /**
+     * Creates a command from a Kafka consumer record.
+     * <p>
+     * The command is created as described in {@link #from(KafkaConsumerRecord)}.
+     * Additionally, gateway information is extracted from the <em>via</em> header of the record.
+     *
+     * @param record The record containing the command.
+     * @return The command.
+     * @throws NullPointerException if record is {@code null}.
+     */
+    public static KafkaBasedCommand fromRoutedCommandRecord(final KafkaConsumerRecord<String, Buffer> record) {
+        final KafkaBasedCommand command = from(record);
+        final String gatewayId = getHeader(record, MessageHelper.APP_PROPERTY_CMD_VIA);
+        if (!Strings.isNullOrEmpty(gatewayId)) {
+            command.setGatewayId(gatewayId);
+        }
+        return command;
+    }
+
     private static String getHeader(final KafkaConsumerRecord<String, Buffer> record, final String header) {
         return RecordUtil.getStringHeaderValue(record.headers(), header);
     }
 
+    /**
+     * Sets the identifier of the gateway this command is to be sent to.
+     *
+     * @param gatewayId The gateway identifier.
+     */
+    public void setGatewayId(final String gatewayId) {
```

Comment:
why do you need to have a setter for the gateway ID? Isn't it determined from the record?

Reply:
Not exclusively - see the comment in `ProtonBasedAdapterInstanceCommandHandler#handleCommandMessage`:
````
// Gateway information set in command handler means a gateway has subscribed for commands for a specific device.
// This information isn't getting set in the message (by the Command Router) and therefore has to be adopted manually here.
````

Reply:
I see. I guess it would make sense to add this information to the JavaDoc here in case other people (or me again in 3 months) are wondering as well :-)

Reply:
Yes, I've added the description of the scenario to the JavaDoc.